Inactive NFS export validation

Via Web-UI and API it is possible to enter, for example, considered invalid `NFS Clients` (Web-UI reference) / `host_string` (model/API reference)  data, resulting in, in this case:

Shares | Host String | Read only | Sync / Async | Actions
-- | -- | -- | -- | --
test-share | 1.! | no | async |  

With the consequent export of:

```
cat /etc/exports
/export/test-share 1.!(rw,async,insecure) 
```

We have in-place a number of validation mechanism intended to filter out, in the above example, what is considered valid host_string contents via the following model field validators:

https://github.com/rockstor/rockstor-core/blob/master/src/rockstor/storageadmin/models/validators.py

representing:

- validate_nfs_host_str()
- validate_nfs_modify_str()
- validate_nfs_sync_choice()

With the 2nd & 3rd validators there is also model `choice` precedence, i.e. the model only accepts ro|rw for modify_str, and async|sync for sync_choice and so we have, in place, redundant validation for those fields. But in the example case of the nfs_host_str (host_str field) used to define the core of this issue, we have at least an inactive validation. And consequently also insufficient API test coverage for this failure.

---

This issue is dependant on NFS test improvements proposed in the following issue:

- NFS tests in need of maintenance #2923

Further to the recent merge of #2925:

- NFS tests in need of maintenance #2923 #2925

We have our proof of inactive validation via the recent test improvements:
```
cd /opt/rockstor/src/rockstor
poetry run django-admin test -v 2 -p test_nfs_export.py
...
test_adv_nfs_get (rockstor.storageadmin.tests.test_nfs_export.NFSExportTests.test_adv_nfs_get) ... ok
test_adv_nfs_post_requests (rockstor.storageadmin.tests.test_nfs_export.NFSExportTests.test_adv_nfs_post_requests) ... ok
test_delete_requests (rockstor.storageadmin.tests.test_nfs_export.NFSExportTests.test_delete_requests) ... ok
test_get (rockstor.storageadmin.tests.test_nfs_export.NFSExportTests.test_get) ... /opt/rockstor/.venv/lib/python3.11/site-packages/rest_framework/pagination.py:207: UnorderedObjectListWarning: Pagination may yield inconsistent results with an unordered object_list: <class 'storageadmin.models.nfs_export_group.NFSExportGroup'> QuerySet.
  paginator = self.django_paginator_class(queryset, page_size)
ok
test_host_str_validator_post (rockstor.storageadmin.tests.test_nfs_export.NFSExportTests.test_host_str_validator_post) ... FAIL
test_host_str_validator_put (rockstor.storageadmin.tests.test_nfs_export.NFSExportTests.test_host_str_validator_put) ... FAIL
test_invalid_get (rockstor.storageadmin.tests.test_nfs_export.NFSExportTests.test_invalid_get) ... ok
test_low_level_error_post (rockstor.storageadmin.tests.test_nfs_export.NFSExportTests.test_low_level_error_post) ... ok
test_low_level_error_put (rockstor.storageadmin.tests.test_nfs_export.NFSExportTests.test_low_level_error_put) ... ok
test_mod_choice_validator_post (rockstor.storageadmin.tests.test_nfs_export.NFSExportTests.test_mod_choice_validator_post) ... FAIL
test_mod_choice_validator_put (rockstor.storageadmin.tests.test_nfs_export.NFSExportTests.test_mod_choice_validator_put) ... FAIL
test_no_nfs_client (rockstor.storageadmin.tests.test_nfs_export.NFSExportTests.test_no_nfs_client) ... ok
test_post (rockstor.storageadmin.tests.test_nfs_export.NFSExportTests.test_post) ... ok
test_post_requests (rockstor.storageadmin.tests.test_nfs_export.NFSExportTests.test_post_requests) ... ok
test_put_requests (rockstor.storageadmin.tests.test_nfs_export.NFSExportTests.test_put_requests) ... ok
test_share_already_exported (rockstor.storageadmin.tests.test_nfs_export.NFSExportTests.test_share_already_exported) ... ok
test_sync_choice_validator_post (rockstor.storageadmin.tests.test_nfs_export.NFSExportTests.test_sync_choice_validator_post) ... FAIL
test_sync_choice_validator_put (rockstor.storageadmin.tests.test_nfs_export.NFSExportTests.test_sync_choice_validator_put) ... FAIL

======================================================================
FAIL: test_host_str_validator_post (rockstor.storageadmin.tests.test_nfs_export.NFSExportTests.test_host_str_validator_post)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/opt/rockstor/src/rockstor/storageadmin/tests/test_nfs_export.py", line 256, in test_host_str_validator_post
    self.assertEqual(
AssertionError: 200 != 500 : {'id': 4, 'exports': [{'id': 4, 'share': 'share2', 'share_id': '22', 'mount': '/export/share2', 'export_group': 4}], 'host_str': '1.!', 'editable': 'ro', 'syncable': 'async', 'mount_security': 'insecure', 'nohide': False, 'enabled': True, 'admin_host': None}

======================================================================
FAIL: test_host_str_validator_put (rockstor.storageadmin.tests.test_nfs_export.NFSExportTests.test_host_str_validator_put)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/opt/rockstor/src/rockstor/storageadmin/tests/test_nfs_export.py", line 274, in test_host_str_validator_put
    self.assertEqual(
AssertionError: 200 != 500 : {'id': 3, 'exports': [{'id': 3, 'share': 'share-nfs', 'share_id': '21', 'mount': '/export/share-nfs', 'export_group': 3}], 'host_str': '*', 'editable': 'rw', 'syncable': 'async', 'mount_security': 'insecure', 'nohide': False, 'enabled': True, 'admin_host': None}

======================================================================
FAIL: test_mod_choice_validator_post (rockstor.storageadmin.tests.test_nfs_export.NFSExportTests.test_mod_choice_validator_post)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/opt/rockstor/src/rockstor/storageadmin/tests/test_nfs_export.py", line 189, in test_mod_choice_validator_post
    self.assertEqual(
AssertionError: 200 != 500 : {'id': 6, 'exports': [{'id': 6, 'share': 'share2', 'share_id': '22', 'mount': '/export/share2', 'export_group': 6}], 'host_str': '*', 'editable': 'rr', 'syncable': 'async', 'mount_security': 'insecure', 'nohide': False, 'enabled': True, 'admin_host': None}

======================================================================
FAIL: test_mod_choice_validator_put (rockstor.storageadmin.tests.test_nfs_export.NFSExportTests.test_mod_choice_validator_put)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/opt/rockstor/src/rockstor/storageadmin/tests/test_nfs_export.py", line 206, in test_mod_choice_validator_put
    self.assertEqual(
AssertionError: 200 != 500 : {'id': 3, 'exports': [{'id': 3, 'share': 'share-nfs', 'share_id': '21', 'mount': '/export/share-nfs', 'export_group': 3}], 'host_str': '*', 'editable': 'rw', 'syncable': 'async', 'mount_security': 'insecure', 'nohide': False, 'enabled': True, 'admin_host': None}

======================================================================
FAIL: test_sync_choice_validator_post (rockstor.storageadmin.tests.test_nfs_export.NFSExportTests.test_sync_choice_validator_post)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/opt/rockstor/src/rockstor/storageadmin/tests/test_nfs_export.py", line 222, in test_sync_choice_validator_post
    self.assertEqual(
AssertionError: 200 != 500 : {'id': 10, 'exports': [{'id': 10, 'share': 'share2', 'share_id': '22', 'mount': '/export/share2', 'export_group': 10}], 'host_str': '*', 'editable': 'ro', 'syncable': 'aaaaa', 'mount_security': 'insecure', 'nohide': False, 'enabled': True, 'admin_host': None}

======================================================================
FAIL: test_sync_choice_validator_put (rockstor.storageadmin.tests.test_nfs_export.NFSExportTests.test_sync_choice_validator_put)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/opt/rockstor/src/rockstor/storageadmin/tests/test_nfs_export.py", line 239, in test_sync_choice_validator_put
    self.assertEqual(
AssertionError: 200 != 500 : {'id': 3, 'exports': [{'id': 3, 'share': 'share-nfs', 'share_id': '21', 'mount': '/export/share-nfs', 'export_group': 3}], 'host_str': '*', 'editable': 'rw', 'syncable': 'async', 'mount_security': 'insecure', 'nohide': False, 'enabled': True, 'admin_host': None}

----------------------------------------------------------------------
Ran 18 tests in 0.827s

FAILED (failures=6)
```

--- a/src/rockstor/storageadmin/models/nfs_export_group.py
+++ b/src/rockstor/storageadmin/models/nfs_export_group.py
@@ -28,9 +28,10 @@ class NFSExportGroup(models.Model):
     INSECURE = "insecure"
 
     """hostname string in /etc/exports"""
-    host_str = models.CharField(max_length=4096, validators=[validators.validate_nfs_host_str])
+    host_str = models.CharField(
+        max_length=4096, validators=[validators.validate_nfs_host_str]
+    )
     """mount options"""
-    """mount read only by default"""
     MODIFY_CHOICES = (
         (READ_ONLY, "ro"),
         (READ_WRITE, "rw"),
@@ -69,3 +70,4 @@ class NFSExportGroup(models.Model):
 
     class Meta:
         app_label = "storageadmin"
+        ordering = ["-id"]

--- a/src/rockstor/storageadmin/views/nfs_exports.py
+++ b/src/rockstor/storageadmin/views/nfs_exports.py
@@ -170,6 +170,7 @@ class NFSExportGroupListView(NFSExportMixin, rfc.GenericView):
 
             cur_exports = list(NFSExport.objects.all())
             eg = NFSExportGroup(**options)
+            eg.clean_fields(exclude="admin_host")
             eg.save()
             for s in shares:
                 mnt_pt = "%s%s" % (settings.MNT_PT, s.name)
@@ -237,6 +238,7 @@ class NFSExportGroupDetailView(NFSExportMixin, rfc.GenericView):
                     s, options["host_str"], request, export_id=int(export_id)
                 )
             NFSExportGroup.objects.filter(id=export_id).update(**options)
+            NFSExportGroup.objects.filter(id=export_id)[0].clean_fields(exclude="admin_host")
             NFSExportGroup.objects.filter(id=export_id)[0].save()
             cur_exports = list(NFSExport.objects.all())
             for e in NFSExport.objects.filter(export_group=eg):

